pyomo/contrib/pynumero/extensions/tests/test_ma27_interface.py

```diff
@@ -0,0 +1,140 @@
+            ma27.set_icntl(1, 0.0)
+
+    def test_set_cntl(self):
+        ma27 = MA27Interface()
+        ma27.set_cntl(1, 1e-8)
+        ma27.set_cntl(3, 1e-12)
+        self.assertAlmostEqual(ma27.get_cntl(1), 1e-8)
+        self.assertAlmostEqual(ma27.get_cntl(3), 1e-12)
+
+    def test_do_symbolic_factorization(self):
+        ma27 = MA27Interface()
+
+        n = 5
+        ne = 7
+        irn = np.array([1,1,2,2,3,3,5], dtype=np.intc)
+        jcn = np.array([1,2,3,5,3,4,5], dtype=np.intc)
+
+        bad_jcn = np.array([1,2,3,5,3,4], dtype=np.intc)
+
+        ma27.do_symbolic_factorization(n, irn, jcn)
+
+        self.assertEqual(ma27.get_info(1), 0)
+        self.assertEqual(ma27.get_info(5), 14) # Min required num. integer words
+        self.assertEqual(ma27.get_info(6), 20) # Min required num. real words
+
+        with self.assertRaisesRegex(AssertionError, 'Dimension mismatch'):
+            ma27.do_symbolic_factorization(n, irn, bad_jcn)
+
+    def test_do_numeric_factorization(self):
+        ma27 = MA27Interface()
+
+        n = 5
+        ne = 7
+        irn = np.array([1,1,2,2,3,3,5], dtype=np.intc)
+        icn = np.array([1,2,3,5,3,4,5], dtype=np.intc)
+        ent = np.array([2.,3.,4.,6.,1.,5.,1.], dtype=np.double)
+        ent_copy = ent.copy()
+        ma27.do_symbolic_factorization(n, irn, icn)
+
+        status = ma27.do_numeric_factorization(irn, icn, n, ent)
+        self.assertEqual(status, 0)
+
+        expected_ent = [2.,3.,4.,6.,1.,5.,1.,]
+        for i in range(ne):
+            self.assertAlmostEqual(ent_copy[i], expected_ent[i])
+    
+        self.assertEqual(ma27.get_info(15), 2) # 2 negative eigenvalues
+        self.assertEqual(ma27.get_info(14), 1) # 1 2x2 pivot
+
+        # Check that we can successfully perform another numeric factorization
+        # with same symbolic factorization
+        ent2 = np.array([1.5, 5.4, 1.2, 6.1, 4.2, 3.3, 2.0], dtype=np.double)
+        status = ma27.do_numeric_factorization(irn, icn, n, ent2)
+        self.assertEqual(status, 0)
+
+        bad_ent = np.array([2.,3.,4.,6.,1.,5.], dtype=np.double)
+        with self.assertRaisesRegex(AssertionError, 'Wrong number of entries'):
+            ma27.do_numeric_factorization(irn, icn, n, bad_ent)
+        with self.assertRaisesRegex(AssertionError, 'Dimension mismatch'):
+            ma27.do_numeric_factorization(irn, icn, n+1, ent)
+
+        # Check that we can successfully perform another symbolic and 
+        # numeric factorization with the same ma27 struct
+        #
+        # n is still 5, ne has changed to 8.
+        irn = np.array([1,1,2,2,3,3,5,1], dtype=np.intc)
+        icn = np.array([1,2,3,5,3,4,5,5], dtype=np.intc)
+        ent = np.array([2.,3.,4.,6.,1.,5.,1.,3.], dtype=np.double)
+        status = ma27.do_symbolic_factorization(n, irn, icn)
+        self.assertEqual(status, 0)
+        status = ma27.do_numeric_factorization(irn, icn, n, ent)
+        self.assertEqual(status, 0)
```

Comment:
Is there any way to test the result of numeric factorization?  Just asserting that the routine returned a `status == 0` seems weak.

Reply:
Agreed that it's a pretty weak test, but I'm not sure what else I could do. Accessing the stored factors (or workspace) arrays would be difficult - it would require me to create a ctypes struct that mimics the memory layout of my `MA27_struct` and to access e.g. the `FACT` array. For the first numeric factorization, I check that two of the `INFO` parameters are what I expect.

The next two factorizations are to make sure I don't have any memory issues when performing numeric factorization then numeric factorization, or numeric factorization then symbolic factorization. I've added checks for the expected number of negative eigenvalues in these matrices, but I'm not sure how worthwhile it would be to try and test beyond the `INFO` arrays.

Reply:
I agree with @Robbybp. The real test of the factorization comes from testing the back solve. If the factorization is not correct, the backsolve will not be correct.